### PR TITLE
Add more `opt_*()` functions

### PR DESCRIPTION
**Pull request by rich-iannone**
These additional `opt_*()` functions provide convenient ways to set multiple table options for typical styling activities.

Here are some examples of the new functions:

```r
library(gt)

tab_1 <-
  exibble %>%
  gt(rowname_col = "row", groupname_col = "group") %>%
  summary_rows(
    groups = "grp_a",
    columns = vars(num, currency),
    fns = list(
      min = ~min(., na.rm = TRUE),
      max = ~max(., na.rm = TRUE)
    )) %>%
  grand_summary_rows(
    columns = vars(currency),
    fns = list(
      total = ~sum(., na.rm = TRUE)
    )) %>%
  tab_source_note(source_note = "This is a source note.") %>%
  tab_footnote(
    footnote = "This is a footnote.",
    locations = cells_body(columns = 1, rows = 1)
  ) %>%
  tab_header(
    title = "The Title of the Table",
    subtitle = "The table's subtitle"
  ) %>%
  tab_stubhead(label = "Top Left")

tab_1

# Table striping is now an add-on
(tab_2 <- tab_1 %>% opt_add_row_striping())

# Using allcaps for labels is common use case for distinguishability
(tab_3 <- tab_2 %>% opt_all_caps())

# Sometimes the additional ink of top/bottom borders may not be desired
(tab_4 <- tab_3 %>% opt_remove_top_bottom_borders())

# The horizontal alignment of the table header can be easily
# modified with `opt_align_table_header()`
(tab_5 <- tab_4 %>% opt_align_table_header("left"))

# There are two options for table lining (one maximal, one minimal)
tab_1 %>% opt_fully_lined()
tab_1 %>% opt_lineless()
```

### CODE REVIEW

File: README.Rmd [hunk 1]
```diff
@@ -66,15 +66,15 @@ sp500 %>%
 
 <p align="center"><img src="man/figures/gt_sp500_table.svg" width="1200px"></p>
 
-There are six datasets included in **gt**: `countrypops`, `sza`, `gtcars`, `sp500`, `pizzaplace`, and `exibble`. All of them are useful for experimenting with the package functions.
+There are six datasets included in **gt**: `countrypops`, `sza`, `gtcars`, `sp500`, `pizzaplace`, and `exibble`. All of them are useful for experimenting with the **gt**'s functions.
```
Comment (jcheng5) [inline README.Rmd]:
```suggestion
There are six datasets included in **gt**: `countrypops`, `sza`, `gtcars`, `sp500`, `pizzaplace`, and `exibble`. All of them are useful for experimenting with **gt**'s functions.
```

File: man/figures/man_info_preset_colorized_1.svg [hunk 2]
```diff
@@ -0,0 +1,5124 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (jcheng5) [inline man/figures/man_info_preset_colorized_1.svg]: Remove this file

File: man/figures/man_location_colors_1.svg [hunk 3]
```diff
@@ -0,0 +1,617 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (jcheng5) [inline man/figures/man_location_colors_1.svg]: Remove this too

File: man/figures/man_opt_remove_table_outline_1.svg [hunk 4]
```diff
@@ -0,0 +1,835 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (jcheng5) [inline man/figures/man_opt_remove_table_outline_1.svg]: The table here is too styled

File: tests/testthat/test-tab_options.R [hunk 5]
```diff
@@ -1463,17 +1463,19 @@ test_that("the row striping options work correctly", {
       selection_text("[class='gt_row gt_left gt_stub gt_striped']"),
     0)
 
+  # TODO: determine why this doesn't work as expected
+
```
Comment (jcheng5) [inline tests/testthat/test-tab_options.R]: ?

File: R/opts.R [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+#' @section Figures:
+#' \if{html}{\figure{man_opt_footnote_marks_1.svg}{options: width=100\%}}
+#'
+#' @family Table Option Functions
+#' @section Function ID:
+#' 9-1
+#'
+#' @export
+opt_footnote_marks <- function(data,
+                               marks = NULL) {
```
Comment (jcheng5) [inline R/opts.R]:
```suggestion
                               marks) {
```

File: R/opts.R [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+
+  # Perform input object validation
+  stop_if_not_gt(data = data)
+
```
Comment (jcheng5) [inline R/opts.R]: Validate `marks` (must be character vector of length >=1, must not be `NULL`)

Comment (jcheng5) [reply R/opts.R]: If the vector is length 1, then also test that it's a valid value

File: R/opts.R [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+#'
+#' @section Figures:
+#' \if{html}{\figure{man_opt_remove_top_bottom_borders_1.svg}{options: width=100\%}}
+#'
+#' @family Table Option Functions
+#' @section Function ID:
+#' 9-5
+#'
+#' @export
+opt_remove_top_bottom_borders <- function(data) {
```
Comment (jcheng5) [inline R/opts.R]: Remove this function--currently it doesn't do anything important that `opt_remove_table_outline` doesn't do

File: NAMESPACE [hunk 9]
```diff
@@ -66,7 +66,15 @@ export(local_image)
...
 export(one_of)
+export(opt_add_row_striping)
+export(opt_add_table_outline)
+export(opt_align_table_header)
+export(opt_all_caps)
 export(opt_footnote_marks)
+export(opt_fully_lined)
+export(opt_lineless)
+export(opt_remove_table_outline)
+export(opt_remove_top_bottom_borders)
```
Comment (jcheng5) [inline NAMESPACE]: Change to:

* `opt_row_striping(TRUE/FALSE)`
* `opt_table_outline(TRUE/FALSE)`
* `opt_align_table_header` (unchanged)
* `opt_all_caps(TRUE/FALSE)`
* `opt_footnote_marks` (unchanged, but add validation to `marks` and no default)
* `opt_table_lines("all"/"none"/"default")`?
* (drop `opt_remove_top_bottom_borders`)

File: R/opts.R (hunk 8 above)
Comment (rich-iannone) [reply R/opts.R]: Now removed.

File: README.Rmd (hunk 1 above)
Comment (rich-iannone) [reply README.Rmd]: Now fixed. Thanks!

File: man/figures/man_info_preset_colorized_1.svg (hunk 2 above)
Comment (rich-iannone) [reply man/figures/man_info_preset_colorized_1.svg]: It's gone.

File: man/figures/man_location_colors_1.svg (hunk 3 above)
Comment (rich-iannone) [reply man/figures/man_location_colors_1.svg]: Also gone.

File: man/figures/man_opt_remove_table_outline_1.svg (hunk 4 above)
Comment (rich-iannone) [reply man/figures/man_opt_remove_table_outline_1.svg]: Now removed.

File: R/opts.R (hunk 6 above)
Comment (rich-iannone) [reply R/opts.R]: Now requires a value.

File: R/opts.R (hunk 7 above)
Comment (rich-iannone) [reply R/opts.R]: Input validation for `marks` has been added (new testthat tests added as well).

File: NAMESPACE (hunk 9 above)
Comment (rich-iannone) [reply NAMESPACE]: All changed now!

File: tests/testthat/test-tab_options.R (hunk 5 above)
Comment (rich-iannone) [reply tests/testthat/test-tab_options.R]: This is a real, but unrelated bug. Using `tbl %>% tab_options(row.striping.include_stub = TRUE)` no ops. Will file an issue for this.

File: R/dt_options.R [hunk 10]
```diff
@@ -169,3 +169,7 @@ dt_options_tbl <-
     "row_striping_include_stub",         FALSE,  "row",              "logical", FALSE,
     "row_striping_include_table_body",   FALSE,  "row",              "logical", FALSE,
   )[-1, ]
+
+dt_options_get_default_value <- function(option) {
+  dt_options_tbl$value[[which(dt_options_tbl$parameter == option)]]
```
Comment (jcheng5) [inline R/dt_options.R]: Throw error if result has 0 rows (i.e. invalid option was passed)

File: R/opts.R [hunk 11]
```diff
@@ -588,103 +563,63 @@ opt_lineless <- function(data) {
...
-#' 9-8
+#' 9-6
 #'
 #' @export
-opt_add_table_outline <- function(data,
-                                  style = "solid",
-                                  width = px(3),
-                                  color = "#D3D3D3") {
+opt_table_outline <- function(data,
+                              active = TRUE,
```
Comment (jcheng5) [inline R/opts.R]: Remove `active = TRUE`, instead the way to remove outline will be `style = "none"`

File: R/dt_options.R (hunk 10 above)
Comment (rich-iannone) [reply R/dt_options.R]: This function now has that validation check.

File: R/opts.R (hunk 11 above)
Comment (rich-iannone) [reply R/opts.R]: Now done. And the other args named `"active"` now have better names.

File: R/opts.R [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+  stop_if_not_gt(data = data)
+
+  # Ensure that all named locations are valid
+  if (!all(locations %in% c("column_labels", "stub", "row_group"))) {
+    stop("The available locations are `column_labels`, `stub`, and `row_group`.",
+         call. = FALSE)
+  }
+
+  # Get vector of `tab_options()` arg names for all specific `locations`
+  options_vec <- table_all_caps_vec(locations)
```
Comment (jcheng5) [inline R/opts.R]: Inline the logic from `table_all_caps_vec` and delete that function; the logic is tightly coupled to this, and you're only using it once.

File: R/opts.R [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+
+  if (all_caps) {
+    values_vec <- rep(c("80%", "bolder", "uppercase"), length(options_vec) / 3)
+    data <- tab_options_multi(data, create_option_value_list(options_vec, values_vec))
```
Comment (jcheng5) [inline R/opts.R]: This implementation is "fragile" in the sense that it's highly dependent on `options_vec` being exactly the right shape (triplets of size, weight, and text_transform, in that order). Instead, let's form `values_vec` based on the suffix of each `options_vec` element (`dplyr::case_when`? `switch`?).

File: R/opts.R [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+  # Perform input object validation
+  stop_if_not_gt(data = data)
+
+  extent <- match.arg(extent)
+
+  # Normalize `extent` values to property values
+  values_vec <- ifelse(extent == "all", "solid", extent)
+
+  # Get vector of `tab_options()` arg names for all specific `locations`
+  options_vec <- table_line_style_vec()
```
Comment (jcheng5) [inline R/opts.R]: Let's inline this as well

File: R/dt_options.R [hunk 15]
```diff
@@ -131,28 +167,107 @@ dt_options_init <- function(data) {
...
+    as.list()
+}
+
+# Do multiple calls of `tab_options()` with an option-value list (`options`)
+tab_options_multi <- function(data, options) {
+  do.call(tab_options, c(list(data = data), options))
+}
+
+# Create vector of all args from `tab_options()` that involve line styles
+table_line_style_vec <- function() {
```
Comment (jcheng5) [inline R/dt_options.R]: Let's introduce a function that does `formals(tab_options) %>% names() %>% setdiff("data") %>% grep(x, ., value = TRUE)`

File: R/opts.R [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+opt_table_lines <- function(data,
+                            extent = c("all", "none", "default")) {
+
+  # Perform input object validation
+  stop_if_not_gt(data = data)
+
+  extent <- match.arg(extent)
+
+  # Normalize `extent` values to property values
+  values_vec <- ifelse(extent == "all", "solid", extent)
```
Comment (jcheng5) [inline R/opts.R]:
```suggestion
  values_vec <- if (extent == "all") "solid" else extent
```

File: R/opts.R [hunk 17]
```diff
@@ -0,0 +1,447 @@
...
+
+  # Get vector of `tab_options()` arg names for all specific `locations`
+  options_vec <- table_line_style_vec()
+
+  if (values_vec %in% c("solid", "none")) {
+    option_value_list <- create_option_value_list(options_vec, values_vec)
+    data <- tab_options_multi(data, option_value_list)
+  } else {
+    option_value_list <- create_default_option_value_list(options_vec)
+    data <- tab_options_multi(data, option_value_list)
```
Comment (jcheng5) [inline R/opts.R]: Pull this line out of the `if`

File: R/opts.R [hunk 18]
```diff
@@ -0,0 +1,447 @@
...
+#' @export
+opt_table_outline <- function(data,
+                              style = "solid",
+                              width = px(3),
+                              color = "#D3D3D3") {
+
+  # Perform input object validation
+  stop_if_not_gt(data = data)
+
+  values_vec <- rep(c(style, width, color), 4)
```
Comment (jcheng5) [inline R/opts.R]: Rewrite this as well using `case_when`

File: R/opts.R [hunk 19]
```diff
@@ -0,0 +1,447 @@
...
+#' This function puts an outline of consistent `style`, `width`, and `color`
+#' around the entire table. It'll write over any existing outside lines so long
+#' as the `width` is larger that of the existing lines. The default value of
+#' `style` (`"solid"`) will draw a solid outline, whereas a value of `"none"`
+#' will remove any present outline.
+#'
+#' @inheritParams fmt_number
+#' @param style,width,color The style, width, and color properties for the table
+#'   outline. By default, these are `"solid"`, `px(3)` (or, `"3px"`), and
+#'   `"#D3D3D3"`.
```
Comment (jcheng5) [inline R/opts.R]: Add a note that `style = "none"` can be used to remove the outline.

File: R/opts.R [hunk 20]
```diff
@@ -0,0 +1,447 @@
...
+#'     footnote = "This is a footnote.",
+#'     locations = cells_body(columns = 1, rows = 1)
+#'   ) %>%
+#'   tab_header(
+#'     title = "The title of the table",
+#'     subtitle = "The table's subtitle"
+#'   ) %>%
+#'   opt_table_outline()
+#'
+#' @section Figures:
```
Comment (jcheng5) [inline R/opts.R]: Maybe an example of `style="none"` as well

File: R/dt_options.R [hunk 21]
```diff
@@ -131,28 +167,107 @@ dt_options_init <- function(data) {
...
+create_default_option_value_list <- function(tab_options_args) {
 
-  dt_options <-
-    data %>%
-    dt_options_get()
+  # Validate the elements of the `tab_options_args` vector
+  validate_tab_options_args(tab_options_args)
 
-  dt_options$value[[which(dt_options$parameter == option)]]
+  tab_options_args %>%
```
Comment (jcheng5) [inline R/dt_options.R]: `lapply(setNames(, tab_option_args), dt_options_get_default_value)`

File: R/opts.R [hunk 22]
```diff
@@ -574,52 +497,14 @@ opt_table_lines <- function(data,
 #'
 #' @export
 opt_table_outline <- function(data,
-                              active = TRUE,
                               style = "solid",
                               width = px(3),
                               color = "#D3D3D3") {
 
   # Perform input object validation
   stop_if_not_gt(data = data)
```
Comment (jcheng5) [inline R/opts.R]: Consider if `style = "none"`, set `width` and `color` to `NULL` (and in the docs, make it clear that "if style is 'none', width and color are ignored")

File: R/opts.R (hunk 12 above)
Comment (rich-iannone) [reply R/opts.R]: Done

File: R/opts.R (hunk 14 above)
Comment (rich-iannone) [reply R/opts.R]: Done.

File: R/dt_options.R (hunk 15 above)
Comment (rich-iannone) [reply R/dt_options.R]: Added `get_tab_options_arg_vec()`. Thanks!

File: R/opts.R (hunk 16 above)
Comment (rich-iannone) [reply R/opts.R]: Changed. Thank you.

File: R/opts.R (hunk 17 above)
Comment (rich-iannone) [reply R/opts.R]: Done.

File: R/opts.R (hunk 18 above)
Comment (rich-iannone) [reply R/opts.R]: Used `case_when()`.

File: R/dt_options.R (hunk 21 above)
Comment (rich-iannone) [reply R/dt_options.R]: Works well. Thanks.

File: R/opts.R (hunk 22 above)
Comment (rich-iannone) [reply R/opts.R]: Implemented now.

File: R/opts.R (hunk 19 above)
Comment (rich-iannone) [reply R/opts.R]: Added that extra documentation.

File: R/opts.R (hunk 20 above)
Comment (rich-iannone) [reply R/opts.R]: Example added. Thanks.

File: R/opts.R (hunk 13 above)
Comment (rich-iannone) [reply R/opts.R]: Fixed now with usage of `dplyr::case_when()`.

File: R/opts.R [hunk 23]
```diff
@@ -290,17 +290,31 @@ opt_all_caps <- function(data,
          call. = FALSE)
   }
 
-  # Get vector of `tab_options()` arg names for all specific `locations`
-  options_vec <- table_all_caps_vec(locations)
+  # Create a regex pattern to obtain arg names for all specific `locations`
+  pattern <-
+    paste0(
+      "^(", paste(locations, collapse = "|"),
+      ")\\.(text_transform|font.size|font.weight)"
```
Comment (jcheng5) [inline R/opts.R]:
```suggestion
      ")\\.(text_transform|font\\.size|font\\.weight)"
```

File: R/opts.R [hunk 24]
```diff
@@ -364,20 +378,18 @@ opt_table_lines <- function(data,
   extent <- match.arg(extent)
 
   # Normalize `extent` values to property values
-  values_vec <- ifelse(extent == "all", "solid", extent)
+  values_vec <- if (extent == "all") "solid" else extent
 
-  # Get vector of `tab_options()` arg names for all specific `locations`
-  options_vec <- table_line_style_vec()
+  # Get vector of `tab_options()` arg names for all table line styles
+  options_vec <- get_tab_options_arg_vec(pattern = ".*style$")
```
Comment (jcheng5) [inline R/opts.R]:
```suggestion
  options_vec <- get_tab_options_arg_vec(pattern = "\\.style$")
```

File: R/opts.R [hunk 25]
```diff
@@ -442,6 +461,98 @@ opt_table_outline <- function(data,
...
+  options_vec <- get_tab_options_arg_vec(pattern = pattern)
+
+  if (is.null(width)) width <- NA_character_
+  if (is.null(color)) color <- NA_character_
+
+  values_vec <-
+    dplyr::case_when(
+      grepl("^table\\.border.*?\\.style$", options_vec) ~ as.character(style),
+      grepl("^table\\.border.*?\\.width$", options_vec) ~ as.character(width),
+      grepl("^table\\.border.*?\\.color$", options_vec) ~ as.character(color)
```
Comment (jcheng5) [inline R/opts.R]: In the above three lines:
`border.*?` -> `border\\..*?`

File: R/opts.R [hunk 26]
```diff
@@ -442,6 +461,98 @@ opt_table_outline <- function(data,
...
 
-  values_vec <- rep(c(style, width, color), 4)
-  tab_options_multi(data, create_option_value_list(table_borders_vec(), values_vec))
+  if (style == "none") {
+    width <- NULL
+    color <- NULL
+  }
+
+  params <-
+    c(style = !is.null(style), width = !is.null(width), color = !is.null(color)) %>%
```
Comment (jcheng5) [inline R/opts.R]:
```r
c(!is.null(style) "style", !is.null(width) "width", is.null(color) "color")
```

File: R/opts.R [hunk 27]
```diff
@@ -442,6 +461,98 @@ opt_table_outline <- function(data,
...
+
+  values_vec <-
+    dplyr::case_when(
+      grepl("^table\\.border.*?\\.style$", options_vec) ~ as.character(style),
+      grepl("^table\\.border.*?\\.width$", options_vec) ~ as.character(width),
+      grepl("^table\\.border.*?\\.color$", options_vec) ~ as.character(color)
+    )
+
+  option_value_list <- create_option_value_list(options_vec, values_vec)
+  tab_options_multi(data, option_value_list)
```
Comment (jcheng5) [inline R/opts.R]: See what happens when e.g. `color = NA` is explicitly passed; if this works, then add unit test

File: R/opts.R [hunk 28]
```diff
@@ -442,6 +461,98 @@ opt_table_outline <- function(data,
...
+}
+
+# Create an option-value list with a vector of arg names from the
+# `tab_options()` function and either one value or n-length values
+# corresponding to those options
+create_option_value_list <- function(tab_options_args, values) {
+
+  # Validate the length of the `values` vector
+  if (length(values) == 1) {
+    values <- rep(values, length(tab_options_args))
```
Comment (jcheng5) [inline R/opts.R]: `rep_len`

File: R/opts.R [hunk 29]
```diff
@@ -442,6 +461,98 @@ opt_table_outline <- function(data,
...
+  }
+
+  dt_options_tbl$value[[which(dt_options_tbl$parameter == option)]]
+}
+
+# Validate any vector of `tab_options()` argument names
+validate_tab_options_args <- function(tab_options_args) {
+
+  tab_options_arg_names <-
+    formals(tab_options) %>% names() %>% base::setdiff("data")
```
Comment (jcheng5) [inline R/opts.R]: `tab_options_arg_names` could be made into a top-level variable. To ensure `tab_options` is defined by the time we get here, use `#' @include tab-options.R`.

File: R/opts.R [hunk 30]
```diff
@@ -442,6 +461,98 @@ opt_table_outline <- function(data,
...
+
+  if (!all(tab_options_args %in% tab_options_arg_names)) {
+    stop("All `tab_options_args` must be valid names.")
+  }
+}
+
+# Do multiple calls of `tab_options()` with an option-value list (`options`)
+tab_options_multi <- function(data, options) {
```
Comment (jcheng5) [inline R/opts.R]: Have this call `validate_tab_options_args(names(options))`, then no need to call that from elsewhere

File: R/opts.R (hunk 23 above)
Comment (rich-iannone) [reply R/opts.R]: Now corrected.

File: R/opts.R (hunk 24 above)
Comment (rich-iannone) [reply R/opts.R]: Now corrected.

File: R/opts.R (hunk 25 above)
Comment (rich-iannone) [reply R/opts.R]: Now changed.

File: R/opts.R (hunk 26 above)
Comment (rich-iannone) [reply R/opts.R]: I couldn't get the suggested line to work (`Error: unexpected string constant in "c(!is.null(style) "style""`). The below does work though:

```
  params <-
    c(
      if (!is.null(style)) "style" else NULL,
      if (!is.null(width)) "width" else NULL,
      if (!is.null(color)) "color" else NULL
    )
```

File: R/opts.R (hunk 28 above)
Comment (rich-iannone) [reply R/opts.R]: Changed. Thanks!

File: R/opts.R (hunk 30 above)
Comment (rich-iannone) [reply R/opts.R]: Done.

File: R/opts.R (hunk 29 above)
Comment (rich-iannone) [reply R/opts.R]: All done. This is great!

File: R/opts.R (hunk 27 above)
Comment (rich-iannone) [reply R/opts.R]: This currently does not work with `NA`:

```
exibble %>% gt() %>% opt_table_outline(color = NA)
```
```
Error: must be a character vector, not a logical vector
Run `rlang::last_error()` to see where the error occurred.
```

```
exibble %>% gt() %>% opt_table_outline(color = NA_character_)
```
```
Error in (function (option, var_name, type)  : 
  Assertion on 'table_border_top_color' failed: Contains missing values (element 1).
```

File: R/opts.R (hunk 26 above)
Comment (jcheng5) [reply R/opts.R]: Yep that's right. The `else NULL` is optional.

**Pull request merged**